src/socket.rs

```diff
@@ -166,3 +168,31 @@ mod linux {
         <u32>::from_be(i)
     }
 }
+
+const EINVAL: i32 = 22;
+pub async fn relay(
+    downstream: &mut tokio::net::TcpStream,
+    upstream: &mut tokio::net::TcpStream,
+) -> Result<(), Error> {
+    #[cfg(all(target_os = "linux"))]
+    {
+        match realm_io::bidi_zero_copy(downstream, upstream).await {
+            Ok(()) => Ok(()),
+            Err(ref e) if e.raw_os_error().map_or(false, |ec| ec == EINVAL) => {
+                match tokio::io::copy_bidirectional(downstream, upstream).await {
+                    Ok(_) => Ok(()),
+                    Err(e) => Err(e),
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    {
+        match tokio::io::copy_bidirectional(downstream, upstream).await {
```

Comment:
I think you can replace this with `map(|_| ())`

Reply:
good suggesttion